imap/jmap_calendar.c

```diff
@@ -1534,23 +1554,43 @@ static int setcalendar_writeprops(jmap_req_t *req,
 
     /* participantIdentities */
     if (!r && json_is_array(props->participant_identities)) {
-        static const char *annot =
-            DAV_ANNOT_NS "<" XML_NS_CALDAV ">calendar-user-address-set";
+        struct caldav_caluseraddr new = CALDAV_CALUSERADDR_INITIALIZER;
+
+        struct caldav_caluseraddr old = CALDAV_CALUSERADDR_INITIALIZER;
+        caldav_caluseraddr_read(mailbox_name(mbox), req->userid, &old);
+
         size_t i;
         json_t *jpid;
         json_array_foreach(props->participant_identities, i, jpid) {
             const char *uri = json_string_value(json_object_get(jpid, "uri"));
             if (!uri) continue;
-            buf_appendcstr(&val, uri);
-            if (i < json_array_size(props->participant_identities)-1)
-                buf_putc(&val, ',');
+            strarray_append(&new.uris, uri);
         }
-        r = annotate_state_writemask(astate, annot, req->userid, &val);
+
+        // Preserve old preferred address, if still available
+        new.pref = strarray_size(&new.uris);
+        if (strarray_size(&old.uris)) {
+            const char *olduri = strarray_nth(&old.uris, old.pref);
+            if (olduri) {
+                int i;
+                for (i = 0; i < strarray_size(&new.uris); i++) {
```

Comment:
Should be able to use strarray_find() is as well.

Reply:
done

Reply:
done